meilisearch-core/src/filters/condition.rs

```diff
@@ -0,0 +1,228 @@
+use super::parser::Rule;
+use crate::{store::Index, DocumentId, MainT};
+use heed::RoTxn;
+use crate::error::Error;
+use meilisearch_schema::{FieldId, Schema};
+use pest::{
+    error::{Error as PestError, ErrorVariant},
+    iterators::Pair,
+};
+use serde_json::{ Value, Number };
+use std::str::FromStr;
+use std::cmp::Ordering;
+
+type ConditionType = Ordering;
+
+/// As Meilisearch does not enforce a schema at indexing, we need to infer type at runtime.
+/// We do it once when the filter is constructed and match every possible types it can be
+/// parsed into.
+///
+/// For now we only support filtering on numbers, booleans, and strings.
+#[derive(Debug)]
+struct ConditionValue<'a> {
+    string: &'a str,
+    boolean: Option<bool>,
+    number: Option<Number>
+}
+
+impl<'a> ConditionValue<'a> {
+    pub fn new(value: &Pair<'a, Rule>) -> Self {
+        let value = match value.as_rule() {
+            | Rule::string
+                | Rule::word => {
+                    let string =  value.as_str();
+                    let boolean = match value.as_str() {
+                        "true" => Some(true),
+                        "false" => Some(false),
+                        _ => None,
+                    };
+                    let number = Number::from_str(value.as_str()).ok();
+                    ConditionValue {
+                        string, boolean, number
+                    }
+                },
+            _ => unreachable!(),
+        };
+        value
+    }
+
+    pub fn as_str(&self) -> &str {
+        self.string.as_ref()
+    }
+
+    pub fn as_num(&self) -> Option<&Number> {
+        self.number.as_ref()
+    }
+
+    pub fn as_bool(&self) -> Option<bool> {
+        self.boolean
+    }
+}
+
+#[derive(Debug)]
+pub struct Condition<'a> {
+    field: FieldId,
+    condition: ConditionType,
+    value: ConditionValue<'a>
+}
+
+fn get_field_value<'a>(schema: &Schema, pair: Pair<'a, Rule>) -> Result<(FieldId, ConditionValue<'a>), Error> {
+    let mut items = pair.into_inner();
+    // lexing ensures that we at least have a key
+    let key = items.next().unwrap();
+    let field = schema
+        .id(key.as_str())
+        .ok_or::<PestError<Rule>>(PestError::new_from_span(
+                ErrorVariant::CustomError {
+                    message: format!(
+                                 "attribute `{}` not found, available attributes are: {}",
+                                 key.as_str(),
+                                 schema.names().collect::<Vec<_>>().join(", ")
+                             ),
+                },
+                key.as_span()))?;
+    let value = ConditionValue::new(&items.next().unwrap());
+    Ok((field, value))
+}
+
+// /!\ undefined behavior with big numbers
```

Comment:
I wouldn't say this is undefined, as it is completely defined, it is a lack of precision that is deterministic.

Reply:
quoting the book:
>Casting from a float to an integer will round the float towards zero
>
>  NOTE: currently this will cause Undefined Behavior if the rounded value cannot be  represented by the target integer type. This includes Inf and NaN. This is a bug and will be fixed.

and 

> Casting from an integer to float will produce the floating point representation of the integer, rounded if necessary (rounding strategy unspecified)

What do you think?

Reply:
Yeah, keep it but please, remove this `/!\` :)